include/exadg/grid/mapping_dof_vector.h

```diff
@@ -105,215 +232,311 @@ class MappingDoFVector
+      }
 
-    std::vector<dealii::types::global_dof_index> dof_indices(fe.dofs_per_cell);
+      // fill grid_coordinates_undeformed
+      MappingTools::fill_grid_coordinates_vector(*mapping_undeformed,
+                                                 grid_coordinates_undeformed,
+                                                 *dof_handler_fe_field);
 
-    for(auto const & cell : dof_handler.active_cell_iterators())
-    {
-      if(not cell->is_artificial())
+      // fill displacement_vector
       {
-        fe_values.reinit(typename dealii::Triangulation<dim>::cell_iterator(cell));
-        cell->get_dof_indices(dof_indices);
-        for(unsigned int i = 0; i < fe_values.n_quadrature_points; ++i)
+        if(displacement.size() != dof_handler_fe_field->n_dofs())
+        {
+          dealii::IndexSet relevant_dofs;
+          dealii::DoFTools::extract_locally_relevant_dofs(*dof_handler_fe_field, relevant_dofs);
+          displacement.reinit(dof_handler_fe_field->locally_owned_dofs(),
+                              relevant_dofs,
+                              dof_handler_fe_field->get_communicator());
+        }
+
+        // Set up dealii::FEValues with FE_Nothing since we only use the geometry information (this
+        // means we need to call fe_values.reinit(cell) with Triangulation::cell_iterator rather
+        // than dealii::DoFHandler::cell_iterator).
+        dealii::FE_Nothing<dim> fe_nothing;
+        dealii::FEValues<dim>   fe_values(*mapping_undeformed,
+                                        fe_nothing,
+                                        dealii::Quadrature<dim>(
+                                          fe_fe_field->base_element(0).get_unit_support_points()),
+                                        dealii::update_quadrature_points);
+
+        std::vector<dealii::types::global_dof_index> dof_indices(fe_fe_field->dofs_per_cell);
+        for(auto const & cell : dof_handler_fe_field->active_cell_iterators())
         {
-          dealii::Point<dim> const point = fe_values.quadrature_point(i);
-          for(unsigned int d = 0; d < dim; ++d)
+          if(cell->is_locally_owned())
           {
-            if(grid_coordinates.get_partitioner()->in_local_range(
-                 dof_indices[component_to_system_index[i][d]]))
+            cell->get_dof_indices(dof_indices);
+
+            fe_values.reinit(typename dealii::Triangulation<dim>::cell_iterator(cell));
+
+            for(unsigned int i = 0; i < fe_fe_field->dofs_per_cell; ++i)
             {
-              grid_coordinates(dof_indices[component_to_system_index[i][d]]) = point[d];
+              unsigned int const d        = fe_fe_field->system_to_component_index(i).first;
+              unsigned int const i_scalar = fe_fe_field->system_to_component_index(i).second;
+
+              dealii::Point<dim> const point = fe_values.quadrature_point(i_scalar);
+
+              displacement(dof_indices[i]) = displacement_function->value(point, d);
```

Comment:
More or less; this code should be considerably more efficient than `interpolate`, because of the de-duplicated support points in the call `fe_values.reinit(cell)`, and because the deal.II function works with some weights for the more general cases of generalized support points. I would leave this code in for now (which is considerably simpler than before), but we could think about using the deal.II function and possibly improving the performance there, see also https://github.com/dealii/dealii/issues/15650 that also lists `interpolate` as poorly implemented.